keps/sig-scheduling/4815-dra-partitionable-devices/README.md

```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
-"flatten" all devices by applying any mixins from their `Includes` fields as
-well as track any capacities consumed from one device by another through its
-`ConsumesCapacityFrom` field. More details on the actual algorithm the
-scheduler follows to make allocation decisions based on the
-`ConsumesCapacityFrom` field can be found in the Design Details section below.
+"flatten" all devices and capacity pools by applying any referenced mixins as
+well as track the capacities consumed by allocated devices. More details
+on the actual algorithm the scheduler follows to make allocation decisions
+based on the capacity pools can be found in the Design Details section below.
 
 ## Design Details
 
 The exact set of proposed API changes can be seen below:
 ```go
 // ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
 type ResourceSliceSpec struct {
-	...
-
-	// DeviceMixins represents a list of device mixins, i.e. a collection of
-	// shared attributes and capacities that an actual device can "include"
-	// to extend the set of attributes and capacities it already defines.
-	//
-	// The main purposes of these mixins is to reduce the memory footprint
-	// of devices since they can reference the mixins provided here rather
-	// than duplicate them.
-	//
-	// The total number of mixins, basic devices, and composite devices must be
-	// less than 128.
-	//
-	// +optional
-	// +listType=atomic
-	DeviceMixins []DeviceMixin `json:"deviceMixins,omitempty"`
+  ...
+
+  // DeviceMixins represents a list of device mixins, i.e. a collection of
+  // shared attributes and capacities that an actual device can "include"
+  // to extend the set of attributes and capacities it already defines.
+  //
+  // The main purposes of these mixins is to reduce the memory footprint
+  // of devices since they can reference the mixins provided here rather
+  // than duplicate them.
+  //
+  // The total number of device mixins, capacity pool mixins, basic devices,
+  // and composite devices must be less than 128.
+  //
+  // +optional
+  // +listType=atomic
+  DeviceMixins []DeviceMixin
+
+  // CapacityPoolMixins represents a list of capacity pool mixins, i.e.
+  // a collection of capacities that a CapacityPool can "include"
+  // to extend the set of capacities it already defines.
+  //
+  // The main purposes of these mixins is to reduce the memory footprint
+  // of capacity pools since they can reference the mixins provided here rather
+  // than duplicate them.
+  //
+  // The total number of device mixins, capacity pool mixins, basic devices,
+  // and composite devices must be less than 128.
+  //
+  // +optional
+  // +listType=atomic
+  CapacityPoolMixins []CapacityPoolMixin
+
+  // SharedCapacityPools defines a list of capacity pools, each of which
+  // has a name and a list of capacities available in the pool.
+  //
+  // The names of the pools must be unique in the ResourceSlice.
+  //
+  // The maximum number of pools is 32.
```

Comment:
So in the NVIDIA case, this limits us to 32 physical GPUs in a ResourceSlice. Which is OK, since you can have more than one slice per node, if we get to very very large nodes.

Reply:
Yeah, it can probably be discussed if these are the right numbers, but in this case users can always spread things across multiple `ResourceSlices`. Each NVIDIA GPU (ref the A100 example) includes a relatively large number of devices, each of which has quite a few annotations and capacities. So I think there might be many reasons for splitting it up into multiple ResourceSlice objects if there are a lot of GPUs.

Reply:
For MIG there are 25 possible MIG devices + 1 full GPU so 26*8=208 total devices for a DGXA100 machine, for example. This is already spilling into more than one slice due to the 128 device limit. With the initial design, my plan was to create one slice per GPU and its MIG (so 8 slices total with 26 devices on each). The one downside being that I need to duplicate all of the mixins in each slice since they cannot span slices.